roofit/roofitcore/inc/TestStatistics/MinuitFcnGrad.h

```diff
@@ -0,0 +1,166 @@
+      LikelihoodWrapperT * /* used only for template deduction */ = static_cast<LikelihoodWrapperT *>(nullptr),
+      LikelihoodGradientWrapperT * /* used only for template deduction */ =
+         static_cast<LikelihoodGradientWrapperT *>(nullptr));
+
+   /// This override should not be used in this class, so it throws.
+   double DoDerivative(const double *x, unsigned int icoord) const override;
+
+   bool syncParameterValuesFromMinuitCalls(const double *x, bool minuit_internal) const;
+
+   // members
+   std::shared_ptr<LikelihoodWrapper> likelihood;
+   std::shared_ptr<LikelihoodGradientWrapper> gradient;
+
+public:
+   mutable std::shared_ptr<WrapperCalculationCleanFlags> calculation_is_clean;
+
+private:
+   mutable std::vector<double> minuit_internal_x_;
+   mutable std::vector<double> minuit_external_x_;
+
+public:
+   mutable bool minuit_internal_roofit_x_mismatch_ = false;
+};
+
+
+/// \param[in] context RooMinimizer that creates and owns this class.
+/// \param[in] parameters The vector of ParameterSettings objects that describe the parameters used in the Minuit
+/// Fitter. Note that these must match the set used in the Fitter used by \p context! It can be passed in from
+/// RooMinimizer with fitter()->Config().ParamsSettings().
+template <typename LikelihoodWrapperT, typename LikelihoodGradientWrapperT>
+MinuitFcnGrad::MinuitFcnGrad(const std::shared_ptr<RooFit::TestStatistics::RooAbsL> &_likelihood, RooMinimizer *context,
+                             std::vector<ROOT::Fit::ParameterSettings> &parameters, bool verbose,
+                             LikelihoodWrapperT * /* value unused */, LikelihoodGradientWrapperT * /* value unused */)
+   : RooAbsMinimizerFcn(RooArgList(*_likelihood->getParameters()), context, verbose), minuit_internal_x_(NDim(), 0),
+     minuit_external_x_(NDim(), 0)
+{
+   synchronizeParameterSettings(parameters, kTRUE, verbose);
+
+   calculation_is_clean = std::make_shared<WrapperCalculationCleanFlags>();
+   likelihood = std::make_shared<LikelihoodWrapperT>(_likelihood, calculation_is_clean /*, _context*/);
+   gradient = std::make_shared<LikelihoodGradientWrapperT>(_likelihood, calculation_is_clean, getNDim(), _context);
+
+   likelihood->synchronizeParameterSettings(parameters);
+   gradient->synchronizeParameterSettings(this, parameters);
+
+   // Note: can be different than RooGradMinimizerFcn, where default options are passed
+   // (ROOT::Math::MinimizerOptions::DefaultStrategy() and ROOT::Math::MinimizerOptions::DefaultErrorDef())
+   likelihood->synchronizeWithMinimizer(ROOT::Math::MinimizerOptions());
+   gradient->synchronizeWithMinimizer(ROOT::Math::MinimizerOptions());
+}
+
+// static function
+template <typename LikelihoodWrapperT, typename LikelihoodGradientWrapperT>
+MinuitFcnGrad *
+MinuitFcnGrad::create(const std::shared_ptr<RooFit::TestStatistics::RooAbsL> &likelihood, RooMinimizer *context,
+                      std::vector<ROOT::Fit::ParameterSettings> &parameters, bool verbose)
+{
```

Comment:
```suggestion
std::unique_ptr<MinuitFcnGrad>
MinuitFcnGrad::create(const std::shared_ptr<RooFit::TestStatistics::RooAbsL> &likelihood, RooMinimizer *context,
                      std::vector<ROOT::Fit::ParameterSettings> &parameters, bool verbose)
{
```

Can we return a smart pointer instead?